<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>ANNOVAR | 变异注释 | HHHH’s Blog</title><meta name="author" content="HHHH"><meta name="copyright" content="HHHH"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="一、简介会得到一系列变异数据，这些变异数据只是告诉我们在基因组的某个位置发生了一段序列的改变，至于这个改变会不会影响生物学功能，我们并不清楚。而注释就是将基因组的序列变异数据转化为我们更关心的生物学功能变化的信息。 Annovar常被用在人类基因组的注释上，其实，它也可以对人类以外的基因组数据进行注释。比如，老鼠基因组的注释上。需要自己进行建立注释信息。 ANNOVAR是一个perl编写的命令行工">
<meta property="og:type" content="article">
<meta property="og:title" content="ANNOVAR | 变异注释">
<meta property="og:url" content="https://0hhhhhh0.github.io/2025/09/16/53/index.html">
<meta property="og:site_name" content="HHHH’s Blog">
<meta property="og:description" content="一、简介会得到一系列变异数据，这些变异数据只是告诉我们在基因组的某个位置发生了一段序列的改变，至于这个改变会不会影响生物学功能，我们并不清楚。而注释就是将基因组的序列变异数据转化为我们更关心的生物学功能变化的信息。 Annovar常被用在人类基因组的注释上，其实，它也可以对人类以外的基因组数据进行注释。比如，老鼠基因组的注释上。需要自己进行建立注释信息。 ANNOVAR是一个perl编写的命令行工">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://0hhhhhh0.github.io/uploads/avatar.png">
<meta property="article:published_time" content="2025-09-16T13:43:00.000Z">
<meta property="article:modified_time" content="2025-09-16T13:53:38.000Z">
<meta property="article:author" content="HHHH">
<meta property="article:tag" content="hexo">
<meta property="article:tag" content="组学，Linux，折腾">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://0hhhhhh0.github.io/uploads/avatar.png"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "ANNOVAR | 变异注释",
  "url": "https://0hhhhhh0.github.io/2025/09/16/53/",
  "image": "https://0hhhhhh0.github.io/uploads/avatar.png",
  "datePublished": "2025-09-16T13:43:00.000Z",
  "dateModified": "2025-09-16T13:53:38.000Z",
  "author": [
    {
      "@type": "Person",
      "name": "HHHH",
      "url": "https://0HHHHHH0.github.io"
    }
  ]
}</script><link rel="shortcut icon" href="/uploads/avatar.png"><link rel="canonical" href="https://0hhhhhh0.github.io/2025/09/16/53/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css?v=5.5.1"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@7.1.0/css/all.min.css"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":true,"top_n_per_article":-1,"unescape":true,"pagination":{"enable":false,"hitsPerPage":8},"languages":{"hits_empty":"No results found for: ${query}","hits_stats":"${hits} articles found"}},
  translate: undefined,
  highlight: {"plugin":"prismjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":true,"highlightMacStyle":false},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Failed',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'null',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.12.0/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'ANNOVAR | 变异注释',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><link rel="stylesheet" href="/css/style.css"><meta name="generator" content="Hexo 7.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="HHHH’s Blog" type="application/atom+xml">
</head><body><div id="loading-box"><div class="scarecrow"><div class="scarecrow__hat"><div class="scarecrow__ribbon"></div></div><div class="scarecrow__head"><div class="scarecrow__eye"></div><div class="scarecrow__eye"></div><div class="scarecrow__mouth"></div><div class="scarecrow__pipe"></div></div><div class="scarecrow__body"><div class="scarecrow__glove scarecrow__glove--l"></div><div class="scarecrow__sleeve scarecrow__sleeve--l"></div><div class="scarecrow__bow"></div><div class="scarecrow__shirt"></div><div class="scarecrow__coat"></div><div class="scarecrow__waistcoat"></div><div class="scarecrow__sleeve scarecrow__sleeve--r"></div><div class="scarecrow__glove scarecrow__glove--r"></div><div class="scarecrow__coattails"></div><div class="scarecrow__pants"></div></div><div class="scarecrow__arms"></div><div class="scarecrow__leg">  </div></div></div><script async="async">(()=>{
  const $loadingBox = document.getElementById('loading-box')
  const $body = document.body
  const preloader = {
    endLoading: () => {
      if ($loadingBox.classList.contains('loaded')) return
      $body.style.overflow = ''
      $loadingBox.classList.add('loaded')
    },
    initLoading: () => {
      $body.style.overflow = 'hidden'
      $loadingBox.classList.remove('loaded')
    }
  }

  preloader.initLoading()

  if (document.readyState === 'complete') {
    preloader.endLoading()
  } else {
    window.addEventListener('load', preloader.endLoading)
    document.addEventListener('DOMContentLoaded', preloader.endLoading)
    // Add timeout protection: force end after 7 seconds
    setTimeout(preloader.endLoading, 7000)
  }

  if (false) {
    btf.addGlobalFn('pjaxSend', preloader.initLoading, 'preloader_init')
    btf.addGlobalFn('pjaxComplete', preloader.endLoading, 'preloader_end')
  }
})()</script><div id="web_bg" style="background-color: #fbfdff;"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/uploads/avatar.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">16</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">1</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">8</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-user"></i><span> About</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/uploads/avatar.png" alt="Logo"><span class="site-name">HHHH’s Blog</span></a><a class="nav-page-title" href="/"><span class="site-name">ANNOVAR | 变异注释</span><span class="site-name"><i class="fa-solid fa-circle-arrow-left"></i><span>  Back to Home</span></span></a></span><div id="menus"><div id="search-button"><span class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> Search</span></span></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-user"></i><span> About</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">ANNOVAR | 变异注释</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2025-09-16T13:43:00.000Z" title="Created 2025-09-16 21:43:00">2025-09-16</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2025-09-16T13:53:38.000Z" title="Updated 2025-09-16 21:53:38">2025-09-16</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E9%87%8D%E6%B5%8B%E5%BA%8F/">重测序</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div><section class="main-hero-waves-area waves-area"><svg class="waves-svg" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" viewBox="0 24 150 28" preserveAspectRatio="none" shape-rendering="auto"><defs><path id="gentle-wave" d="M -160 44 c 30 0 58 -18 88 -18 s 58 18 88 18 s 58 -18 88 -18 s 58 18 88 18 v 44 h -352 Z"></path></defs><g class="parallax"><use href="#gentle-wave" x="48" y="0"></use><use href="#gentle-wave" x="48" y="3"></use><use href="#gentle-wave" x="48" y="5"></use><use href="#gentle-wave" x="48" y="7"></use></g></svg></section></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><div class="ai-summary"><div class="ai-explanation" style="display: block;" data-summary="这里是清羽AI，这篇文章介绍了ANNOVAR变异注释工具的核心功能和使用方法。ANNOVAR是一个基于Perl的命令行工具，用于将基因组变异数据转化为生物学功能变化信息，支持人类及其他物种（如小鼠）的基因组注释。文章详细讲解了三种注释类型：基于基因的注释（确定突变位点的位置和功能影响）、基于区域的注释（识别特定基因组区域的变异）以及基于变异数据库的注释（提供population frequency等信息）。此外，文章还提供了ANNOVAR的下载安装指南、文件结构说明、数据库下载方法以及具体的注释操作步骤，包括如何对人类基因组和非人类基因组进行注释。">AI正在绞尽脑汁想思路ING···</div><div class="ai-title"> <div class="ai-title-left"> <i class="fa-brands fa-slack"></i><div class="ai-title-text">AI摘要</div></div><div class="ai-tag" id="ai-tag">deepseek-ai/DeepSeek-V3</div></div></div><h2 id="一、简介"><a href="#一、简介" class="headerlink" title="一、简介"></a>一、简介</h2><p>会得到一系列变异数据，这些变异数据只是告诉我们在基因组的某个位置发生了一段序列的改变，至于这个改变会不会影响生物学功能，我们并不清楚。而注释就是将基因组的序列变异数据转化为我们更关心的生物学功能变化的信息。</p>
<p>Annovar常被用在人类基因组的注释上，其实，它也可以对人类以外的基因组数据进行注释。比如，老鼠基因组的注释上。需要自己进行建立注释信息。</p>
<p>ANNOVAR是一个perl编写的命令行工具，能在安装了perl解释器的多种操作系统上执行。允许多种输入文件格式，包括最常被使用的VCF格式。输出文件也有多种格式，包括注释过的VCF文件、用tab或者逗号分隔的text文件。</p>
<p>ANNOVAR能快速注释遗传变异并预测其功能。类似的variants注释软件还有 VEP, snpEff, VAAST, AnnTools等等.</p>
<hr>
<p>ANNOVAR 注释变异可以分成有基于基因、基于染色体区间和变异数据等三种类型. 这三种注释分别针对于每一个variant的不同方面：</p>
<p>基于基因的注释（gene-based annotation）<br>注释结果为突变位点位于基因的相对位置，是否改变氨基酸编码，确定受影响的氨基酸，获得变异位点的HGVS命名方式，揭示variant与已知基因直接的关系以及对其产生的功能性影响。可灵活使用RefSeq genes, UCSC genes, ENSEMBL genes, GENCODE genes或许多其他基因定义系统。</p>
<p>基于染色体区间的注释（region-based annotation）<br>识别特定基因组区域的变异，例如，44个物种中的保守区域，预测的转录因子结合位点, segmental duplication regions, GWAS hits, ChIP-Seq peaks, RNA-Seq peaks等等许多其他的在基因组区间的注释；</p>
<p>变异数据库的注释 / 基于过滤的注释（ filter-based annotation ）<br>则给出这个variant的一系列信息，如： population frequency in different populations 和various types of variant-deleteriousness prediction scores, 这些可被用来过滤掉一些公共的及 probably（大概,肯定的成分较大,是most likely） nondeleterious variants. 包括Clinvar, dbSNP, Cosmic, ExAC, gnomAD等等，突变数据库整理可参考从 vcf 文件准备 ANNOVAR 数据库。鉴定特定数据库中记录的变异，例如，该变异位点是否在dbSNP中有报道，在千人基因组计划中的等位基因频率如何等等。</p>
<p>ANNOVAR 数据库文件实际上为特定格式的文本文件，其数据库文件命名规则为: ${path_database}/${buildver}_${database_name}.txt</p>
<p>ANNOVAR 所有注释结果都在 vcf 文件 INFO 列添加key-value</p>
<h2 id="二、ANNOVAR的下载"><a href="#二、ANNOVAR的下载" class="headerlink" title="二、ANNOVAR的下载"></a>二、ANNOVAR的下载</h2><p>下载地址</p>
<p>填写登记表，下载ANNOVAR软件（<a target="_blank" rel="noopener" href="http://annovar.openbio/">http://annovar.openbio</a> informatics.org/），解压文件</p>
<p>tar xvfz annovar.latest.tar.gz<br>解压后生成annovar文件夹，里面有6个perl脚本程序和2个文件夹。</p>
<h3 id="ANNOVAR-的文件结构"><a href="#ANNOVAR-的文件结构" class="headerlink" title="ANNOVAR 的文件结构"></a>ANNOVAR 的文件结构</h3><p>|– annotate_variation.pl              # 主程序，功能包括下载数据库，三种不同的注释<br>|– coding_change.pl                   # 可用来推断蛋白质序列<br>|– convert2annovar.pl                 # 将多种格式转为.avinput的程序<br>|– retrieve_seq_from_fasta.pl         # 用于自行建立其他物种的转录本<br>|– table_annovar.pl                   # 注释程序，可一次性完成三种类型的注释<br>|– variants_reduction.pl              # 可用来更灵活地定制过滤注释流程<br>|– example                            # 存放示例文件<br>|– humandb                            # 人类注释数据库<br>自带了humandb是已经建立好的hg19或者GRCh37等常用的数据库，可用于人的注释。 如果要进行其他注释，需要使用 -downdb命令下载数据库到 humandb/ 目录。</p>
<p>三、注释<br>perl retrieve_seq_from_fasta.pl<br>  –format refGene<br>  –seqfile zunla.fasta  zunla_refGene.txt<br>  –out zunla_refGeneMrna.fa</p>
<ol>
<li>用ANNOVAR注释人类基因组variants信息<br>1.1 下载所有需要的注释信息库<br>对于基于基因的注释的数据库已经在下好的 ANNOVAR package中了。如果要进行其他注释，需要按以下命令下载数据库到 ‘humandb/’ 目录里：<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">perl annotate_variation.pl --downdb --buildver hg19 cytoBand humandb/
perl annotate_variation.pl --downdb --webfrom annovar --buildver hg19 1000g2014oct humandb/
perl annotate_variation.pl --downdb --webfrom annovar --buildver hg19 exac03 humandb/
perl annotate_variation.pl --downdb --webfrom annovar --buildver hg19 ljb26_all humandb/
perl annotate_variation.pl --downdb --webfrom annovar --buildver hg19 clinvar_20140929 humandb/
perl annotate_variation.pl --downdb --webfrom annovar --buildver hg19 snp138 humandb/

# -buildver: 基因组对应版本
# -webfrom annovar: 从annovar库里下载;如果annovar库中没有，则不用写该选项，会从UCSC中下载
# refGene: 数据库名称
# humandb/: 下载至该目录<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre></li>
</ol>
<p>这里下载的是几个通常用到的数据库：<br>1）cytoBand 是每个细胞间band（cytogenetic band）的染色体坐标信息 ,<br>2）1000g2014oct<br>　　　　for alternative allele frequency in the 1000 Genomes Project (version October 2014),<br>　　　　是2014年10版，1000基因组项目（和ExAV 外显子集合联合一样，是公开、开放的数据库）<br>　　　　里面供选择的等位基因频率信息<br>3）exac03<br>　　　　for the variants reported in the Exome Aggregation Consortium (version 0.3),<br>　　　　是0.3版外显子集合联合中报道过的variants.<br>4）ljb26_all<br>　　　　for various functional deleteriousness prediction scores from the dbNSFP database (version 2.6),<br>　　　　 dbNSFP: A Lightweight Database of Human NonsynonymousSNPs and TheirFunctionalPredictions on ResearchGate<br>5）clinvar_20140929<br>　　　　for the variants reported in the ClinVar database (version 20140929)<br>　　　　ClinVar是美国国家生物技术信息中心（NCBI）于2012年11月宣布、2013年4月正式启动的公共、免费数据库。作为核心数据库，ClinVar数据库整合了十多个不同类型数据库、通过标准的命名法来描述疾病，同时支持科研人员将数据下载到本地中，开展更为个性化的研究。在遗传变异和临床表型方面，NCBI和不同的研究组已经建立了各种各样的数据库，数据信息相对比较分散，ClinVar数据库的目的在于整合这些分散的数据、将变异、临床表型、实证数据以及功能注解与分析等四个方面的信息，通过专家评审，逐步形成一个标准的、可信的、稳定的遗传变异-临床表型相关的数据库。<br>6）snp138<br>　　　　for the dbSNP database (version 138).</p>
<p>注意：<br>　　(i) 第一个命令中不包含 –webfrom annovar 选项, 因此是从the UCSC Genome Browser annotation database下载文件的；<br>　　(ii) –buildver hg19 选项是针对hg19这一版的基因组的；<br>　　(iii) 运行上面命令后，在 humandb/ 目录下会多几个以 hg19为前缀的文件。</p>
<p>1.2 table_annovar.pl注释variants<br>输入下列命令，用之前下载好的注释数据库来注释vcf格式文件中的variants。ANNOVAR 所有注释结果都在 vcf 文件 INFO 列添加key-value。输出的csv文件将包含输入的5列主要信息以及各个数据库里的注释。</p>
<p>perl table_annovar.pl  humandb/ –outfile final –buildver hg19<br>  –protocol refGene,cytoBand,1000g2014oct_eur,1000g2014oct_afr,exac03,ljb26_all,clinvar_20140929,snp138<br>  –operation g,r,f,f,f,f,f,f<br>  –vcfinput<br> 参考输入的vcf文件的名称<br>　　–protocol 选项后跟注释来源数据库的准确名称<br>　　–operation选项后跟注释的类型:<br>　　　　g 表示基于基因的注释（gene-based annotation）；<br>　　　　r 表示基于区域的注释（region-based annotation）；<br>　　　　f 表示基于筛选子的注释（ filter-based annotation）；<br>　　–outfile 选项是指定输出文件的前缀</p>
<p>关键步骤：<br>　　1、确保注释数据库的名称正确并且是按你想要在输出文件中显示的顺序排列的；<br>　　2、确保 –operation指定的注释类型顺序和–protocol指定的数据库顺序是一致的；<br>　　3、确保每个protocal名称或注释类型之间只有一个逗号，并且没有空白。</p>
<p>final.hg19_multianno.vcf.输出文件应该是以个VCF格式文件，INFO那列以 key=value 形式、 ;分割成几个小区域. eg:Func.refGene=intronic;Gene.refGene=SAMD11. 每个键值对代表一个ANNOVAR注释信息。输出文件可以用为VCF格式文件设计的基因分析软件进一步处理。</p>
<p>final.hg19_multianno.txt. 每一行代表一个variant 。用tab分隔，多余列为加上的注释信息，顺序按 –protocol选项所设定的注释类型argument。</p>
<ol start="2">
<li>用Annovar注释人类以外的基因组（Gene-based annotation）<br>Annovar常被用在人类基因组的注释上，其实，它也可以对人类以外的基因组数据进行注释。annovar一般只包含人类基因组注释数据库，其他的物种需要自己进行建立注释信息。</li>
</ol>
<p>2.1 以注释小鼠基因组为例<br>一般如果你想看是否有某种物种,如小鼠mm9的注释库时，命令行运行</p>
<p>perl annotate_variation.pl -builder mm9 -downdb avdblist -webfrom annovar ./<br>会生成一个mm9开头的文件，里面包含小鼠mm9有多少注释数据库，然后自己可以构建一个mousedb数据库 先在annovar文件夹里面创建mousedb文件夹（名字可自取），命令mkdir mousedb 然后使用annovar文件夹下的perl程序annotate_variation.pl</p>
<p>perl annotate_variation.pl -downdb -buildver mm9 -webfrom annovar refGene mousedb/<br>这个命令能实现的是帮忙下载mm9的refGene的文件，保存在mousedb文件下，自动解压后文件名为mm9_refGene.txt。 然后程序会提示使用以下两个命令继续建库</p>
<p>annotate_variation.pl –buildver mm9 –downdb seq mousedb/mm9_seq<br>retrieve_seq_from_fasta.pl mousedb/mm9_refGene.txt -seqdir mousedb/mm9_seq -format refGene -outfile mousedb/mm9_refGeneMrna.fa<br>同样在annovar文件下运行这两个perl程序</p>
<p>perl annotate_variation.pl –buildver mm9 –downdb seq mousedb/mm9_seq<br>通过这个命令，会在mousedb下创建文件夹mm9_seq，并且在里面下载mm9的基因组文件chromFa.tar.gz，perl程序帮忙解压后是按染色体分开的fasta格式文件。 然后继续运行perl程序</p>
<p>perl retrieve_seq_from_fasta.pl mousedb/mm9_refGene.txt -seqdir mousedb/mm9_seq -format refGene -outfile mousedb/mm9_refGeneMrna.fa<br>该程序会会在mousedb下创建mm9_refGeneMrna.fa文件，是根据mm9_refGene.txt的信息，重新构建成的老鼠转录表达基因fasta格式文件 这样老鼠mm9 annovar gene based注释库就弄好了 以文本文件test.input为案例进行测试 生成test.input的txt格式文件，根据annovar官网介绍，只要这最基本的五列信息就可以进行注释，五列分别染色体名称，染色体上的位置，染色体上的位置，参考基因组碱基，变异碱基。</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">1       19215217        19215217        T       C
1       33803084        33803084        A       G
1       33803198        33803198        A       G
1       37499237        37499237        T       C
1       37499238        37499238        T       C
1       37500003        37500003        T       C
1       43826936        43826936        T       C
1       58853960        58853960        A       G
1       58854487        58854487        A       G
1       60436865        60436865        T       C<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>然后使用perl程序进行gene based的注释</p>
<p>perl annotate_variation.pl -out test -build mm9 test.input mousedb<br>注释后会生成test.variant_function，test.exonic_variant_function和test.log文件，前两个即为所需要的文件。用这个例子输出test.exonic_variant_function文件输出为空文件，因为这些位点没有在exonic区域的，所以没有结果。如果有位点在exonic中，则在test.exonic_variant_function中会更具体的描述为同义突变还是非同义突变</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">intronic        Tfap2b  1       19215217        19215217        T       C
UTR3            Bag2    1       33803084        33803084        A       G
UTR3            Bag2    1       33803198        33803198        A       G
UTR3            Mgat4a  1       37499237        37499237        T       C
UTR3            Mgat4a  1       37499238        37499238        T       C
UTR3            Mgat4a  1       37500003        37500003        T       C
intronic        Uxs1    1       43826936        43826936        T       C
intronic        Casp8   1       58853960        58853960        A       G
intronic        Casp8   1       58854487        58854487        A       G
intronic        Cyp20a1 1       60436865        60436865        T       C<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>2.2 以注释大猩猩基因组（with the genome build identifier as panTro2.）为例。<br>对于gene-based annotation， ANNOVAR需要 genePred format 的 gene definition file和 FASTA format 的 transcript sequence file；</p>
<p>第一步：输入以下命令，下载大猩猩基因组定义文件（ gene definition file）及序列的 FASTA 文件到chimpdb/目录</p>
<p>perl annotate_variation.pl –downdb –buildver panTro2 gene chimpdb/<br>perl annotate_variation.pl –downdb –buildver panTro2 seq chimpdb/panTro2_seq<br>运行过程中，出现下列提示</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">---------------------------ADDITIONAL PROCEDURE---------------------------
--------------------------------------------------------------------------
NOTICE: the FASTA file for the genome is not available to download but can be generated by the ANNOVAR software.
PLEASE RUN THE FOLLOWING TWO COMMANDS CONSECUTIVELY TO GENERATE THE FASTA FILES (you may need to change -seqdir to -seqfile for some genomes):

    annotate_variation.pl --buildver panTro2 --downdb seq chimpdb/panTro2_seq
    retrieve_seq_from_fasta.pl chimpdb/panTro2_refGene.txt -seqdir chimpdb/panTro2_seq -format refGene -outfile chimpdb/panTro2_refGeneMrna.fa<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>


<p>第二步：注意ANNOVAR数据库中只包含人类基因组已建好的转录本，不包含其他物种的。故需要按以下命令自行建立对应物种的transcript FASTA file</p>
<p>perl retrieve_seq_from_fasta.pl chimpdb/panTro2_refGene.txt<br>  –seqdir chimpdb/panTro2_seq<br>  –format refGene<br>  –outfile chimpdb/panTro2_refGeneMrna.fa<br>–seqdir说明下载的序列文件的所在目录；<br>–format说明 gene definition file的格式.；<br>–outfile 指定输出mRNA 序列文件的名称；</p>
<p>关键：跟在–outfile后的输出文件名应该是 _refGeneMrna.fa这种形式，否则下一步找不到正确的 transcript FASTA sequence file.</p>
<p>第三步：注释variants，with the chimpanzee gene annotation:</p>
<p>perl table_annovar.pl  chimpdb/<br>  –vcfinput<br>  –outfile final<br>  –buildver panTro2<br>  –protocol refGene<br>  –operation g<br> input VCF file；<br>chimpdb/ directory of the downloaded data；</p>
<p>第四步：输出结果文件核对。 final.panTro2_multianno.txt file. The gene annotation for chimpanzee is added after the input variants.</p>
<p>关键：如果没有现成可用的gene definition file ，可以将基因预测工具产生的 GFF3 or GTF 文件转换成 gene definition file.</p>
<div style="height:3px;background:linear-gradient(90deg,#ff6c6c,#73aaff);margin:2rem 0;border-radius:2px;"></div>


<h2 id="三、构建自定义注释库"><a href="#三、构建自定义注释库" class="headerlink" title="三、构建自定义注释库"></a>三、构建自定义注释库</h2><p>ANNOVAR可以从服务器下载注释库，但是主要针对人类基因组，那么需要分析、注释其它的物种测序结果，怎么办呢，需要自建注释库。</p>
<p>1 以注释长牡蛎基因组为例</p>
<h3 id="第一步：准备工作"><a href="#第一步：准备工作" class="headerlink" title="第一步：准备工作"></a>第一步：准备工作</h3><p>首先，需要参考基因组的序列和注释文件，这里是名为Cg长牡蛎</p>
<p>GCF_902806645.1_cgigas_uk_roslin_v1_genomic.fna      # 长牡蛎 的参考基因组序列<br>/home/wanglab/Ywh/adductor_muscle_scar/Crassostrea_gigas.cgigas_uk_roslin_v1.61.gtf       # 长牡蛎的注释文件，我这里是gtf<br>ANNOVAR 建库需要 genePred 文件，因而需要准备几个转换 gtf 到 genePred 格式的软件</p>
<p>gffread           # gff3 to gtf<br>gtfToGenePred     # gtf to genePred<br>gffread的下载地址，需要自行编译。编译过程会自行下载<a target="_blank" rel="noopener" href="https://github.com/gpertea/gclib%EF%BC%8C%E4%B9%9F%E5%8F%AF%E4%BB%A5%E9%A2%84%E5%85%88%E4%B8%8B%E8%BD%BD%EF%BC%8C%E7%84%B6%E5%90%8E">https://github.com/gpertea/gclib，也可以预先下载，然后</a> <code>make</code> 。</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">git clone https://github.com/gpertea/gffread
cd gffread
make release<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span></span></code></pre>

<p>推荐直接conda安装</p>
<p> <code>conda install gffread</code><br>gtfToGenePred的下载地址，已经编译好了，下载直接使用，他属于UCSC工具包，因而在conda环境安装gtfToGenePred的命令为</p>
<p> <code>conda install ucsc-gtftogenepred</code><br>conda直接安装gtftogenepred没成功，需要加上UCSC前缀。</p>
<h3 id="第二步：建立注释库"><a href="#第二步：建立注释库" class="headerlink" title="第二步：建立注释库"></a>第二步：建立注释库</h3><p>创建辣椒注释库的文件夹Cgigasdb，然后进去</p>
<p>mkdir Cgigasdb<br>cd Cgigasdb<br>转换gff3 为 gtf，推荐使用GTF格式，因为有些GFF3格式文件转换可能不正确</p>
<p>gffread Cgigas.gff -T -o Cgigas.gtf<br>转换gtf 为 GenePred</p>
<p>gtfToGenePred -genePredExt “/home/wanglab/Ywh/adductor_muscle_scar/ReSeq_Data/DZQD2023092909/annovar/Cgigasdb/GCF_902806645.1_cgigas_uk_roslin_v1_genomic.gtf” Cg_refGene.txt<br>建立注释库</p>
<p>perl ../retrieve_seq_from_fasta.pl –format refGene –seqfile GCF_902806645.1_cgigas_uk_roslin_v1_genomic.fna –outfile Cg_refGeneMrna.fa Cg_refGene.txt<br>这样我们的建库就完成了，获得两个重要的文件</p>
<p>Cgigas_refGeneMrna.fa<br>Cgigas_refGene.txt<br>强调：关于文件的命名，按照常规逻辑，这两个文件肯定不能随便命名，不然annovar无法识别！摸索了一下，前缀就是下面build参数使用的名字，这里就是Cgigas，下面注释就要使用“-build Cgigas”这个参数，对于基于基因注释的txt文件命名就是refGene，连起来就是 Cgigas_refGene.txt。而fa文件前缀一样，后面有稍稍差别为refGeneMrna，连起来就是Cgigas_refGeneMrna.fa。</p>
<h3 id="第三步：转换需要注释的vcf文件"><a href="#第三步：转换需要注释的vcf文件" class="headerlink" title="第三步：转换需要注释的vcf文件"></a>第三步：转换需要注释的vcf文件</h3><p>BB3.HC.vcf是HaplotypeCaller得到的vcf文件，转换成适用annovar的文件格式。得到的BB3.avinput就是我们需要注释的文件</p>
<p>perl convert2annovar.pl -format vcf4 BB3.HC.vcf &gt; BB3.avinput</p>
<pre class="line-numbers language-C" data-language="C"><code class="language-C"># 将您的TSV转换为Annovar的avinput格式
awk 'NR&gt;1 {print $1"\t"$2"\t"$2"\t"$3"\t"$4}' /home/wanglab/Ywh/adductor_muscle_scar/Variants/filtered_variants_thr0p7_20250902_162128.tsv &gt; filtered_variants_thr0p7_20250902_162128.avinput<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span></span></code></pre>

<p>ANNOVAR主要使用convert2annovar.pl程序进行转换，转换后文件是精简过的，主要包含前面提到的5列内容，如果要将原格式的文件的所有内容都包含在转换后的.avinput文件中，可以使用-includeinfo参数；如果需要分开每个sample输出单一的.avinput文件，可以使用-allsample参数，等等。</p>
<p>ANNOVAR还主要支持以下格式转换：</p>
<p>SAMtools pileup format<br>Complete Genomics format<br>GFF3-SOLiD calling format<br>SOAPsnp calling format<br>MAQ calling format<br>CASAVA calling format</p>
<h2 id="第四步：annotate-variation注释"><a href="#第四步：annotate-variation注释" class="headerlink" title="第四步：annotate_variation注释"></a>第四步：annotate_variation注释</h2><pre class="line-numbers language-C" data-language="C"><code class="language-C"># 一步完成注释和表格输出
perl table_annovar.pl filtered_variants_thr0p7_20250902_162128.avinput Cgigasdb/ \
   -buildver Cg \
   -out filtered_variants_thr0p7_20250902_162128 \
   -remove \
   -protocol refGene \
   -operation g \
   -nastring . \
   -csvout<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>-geneanno -dbtype refGene是默认值，可以省略，那么命令也可以写成</p>
<p>得到结果文件：</p>
<p>BB3.exonic_variant_function     # 外显子区域突变的功能、类型等<br>BB3.variant_function            # 突变的基因及位置<br>BB3.log                         # 日志文件<br>第五步：table_annovar.pl注释</p>
<p>table_annovar.pl是ANNOVAR多个脚本的封装，可以一次性完成三种类型的注释。</p>
<p>我这里只有regGene类型的注释库，那么注释命令为</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">perl table_annovar.pl  
  BB3.avinput pepperdb/  
  -buildver zunla               # 使用zunla注释库
  -out BB3 
  -remove                       # 清除所有临时文件
  -protocol refGene             # 注释库类型为refgene
  -operation g                  # 操作子为g
  -nastring .                   # 缺省值用“.”代替
  -csvout                       # 输出csv文件<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>table_annovar.pl也可以不经过转换，直接对vcf文件进行注释，添加-vcfinput参数就行，注释的内容将会放在vcf文件的“INFO”那一栏。但是需要注意的是不能和-csvout参数同时使用，命令如下</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">perl table_annovar.pl 
  BB3.HC.vcf 
  pepperdb/  
  -buildver zunla 
  -out BB3 
  -protocol refGene 
  -operation g 
  -nastring .  
  -vcfinput 
  -polish<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>所以，两种输入格式</p>
<p>VCF文件：用 -vcfinput指定<br>avinput<br>每行代表一个位点<br>前5列依次为：chromosome, start position, end position, the reference nucleotides, the observed nucleotides<br>reference nucleotides：不知道时可设置为0<br>observed nucleotides: insertion,deletion,block subsititution可用-表示<br>其余列：可有可无，如果有，在输出文件中会原样输出。<br>more BB3.avinput</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">1       948921  948921  T       C       comments: rs15842, a SNP in 5' UTR of ISG15
1       13211293        13211294        TC      -       comments: rs59770105, a 2-bp deletion
1       11403596        11403596        -       AT      comments: rs35561142, a 2-bp insertion
1       105492231       105492231       A       ATAAA   comments: rs10552169, a block substitution
1       67705958        67705958        G       A       comments: rs11209026 (R381Q), a SNP in IL23R associated <span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span></span></code></pre>


<p>cat example/ex1.avinput</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">1   948921  948921  T   C   comments: rs15842, a SNP in 5' UTR of ISG15
1   1404001 1404001 G   T   comments: rs149123833, a SNP in 3' UTR of ATAD3C
1   5935162 5935162 A   T   comments: rs1287637, a splice site variant in NPHP4
1   162736463   162736463   C   T   comments: rs1000050, a SNP in Illumina SNP arrays
1   84875173    84875173    C   T   comments: rs6576700 or SNP_A-1780419, a SNP in Affymetrix SNP arrays
1   13211293    13211294    TC  -   comments: rs59770105, a 2-bp deletion
1   11403596    11403596    -   AT  comments: rs35561142, a 2-bp insertion
1   105492231   105492231   A   ATAAA   comments: rs10552169, a block substitution
1   67705958    67705958    G   A   comments: rs11209026 (R381Q), a SNP in IL23R associated with Crohn's disease
2   234183368   234183368   A   G   comments: rs2241880 (T300A), a SNP in the ATG16L1 associated with Crohn's disease<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>该格式每列以tab分割，最重要的地方为前5列，分别是:</p>
<p>染色体(Chromosome)<br>起始位置(Start)<br>结束位置(End)<br>参考等位基因(Reference Allele)<br>替代等位基因(Alternative Allele)<br>剩下为注释部分（可选）。<br>ANNOVAR主要也是依靠这5处信息对数据库进行比对，进而注释变异。</p>
<p>报错：</p>
<p>chmod 777 table_annovar.pl<br>问题依旧！</p>
<p>没办法，只能跑去table_annovar.pl脚本的444行看一下具体代码</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">1.  #generate gene anno
2.  my $sc;
3.  $sc = $SC_PREFIX .  "annotate_variation.pl -geneanno -buildver $buildver -dbtype $protocol -outfile $tempfile.$protocol -exonsort -nofirstcodondel $queryfile $dbloc";  #20191010: add -nofirstcodondel
4.  $argument and $sc .=  " $argument";
5.  $intronhgvs and $sc .=  " -splicing_threshold $intronhgvs";

7.  if  ($thread)  {
8.  $sc .=  " -thread $thread";
9.  }
10.  if  ($maxgenethread)  {
11.  $sc .=  " -maxgenethread $maxgenethread";
12.  }

14.  print STDERR "\nNOTICE: Running with system command \n";
15.  system ($sc)  and  die  "Error running system command: \n";<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>


<p>444行的内容为system ($sc) and die “Error running system command: \n”;<br>看得出来，这里执行了system系统命令，往上查找变量$sc的值，发现这里调用了annotate_variation.pl脚本，同样</p>
<p>chmod 777 annotate_variation.pl<br>为了避免别的脚本权限问题，干脆全部777</p>
<p>chmod 777 *<br>成功解决权限问题。</p>
<p>注释完的结果合并到原文件</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup"># 步骤1: 为原始文件添加行号
awk 'NR==1 {print "LineNumber\t"$0} NR&gt;1 {print NR-1"\t"$0}' /home/wanglab/Ywh/adductor_muscle_scar/Variants/filtered_variants_thr0p7_20250902_162128.tsv &gt; original_with_linenum.tsv

# 步骤2: 为注释文件添加行号
awk 'NR==1 {print "LineNumber\t"$0} NR&gt;1 {print NR-1"\t"$0}' filtered_variants_thr0p7_20250902_162128.Cg_multianno.csv &gt; anno_with_linenum.tsv

# 步骤3: 基于行号合并
join -t $'\t' -1 1 -2 1 original_with_linenum.tsv anno_with_linenum.tsv &gt; merged_result.tsv

# 步骤4: 清理临时文件
rm original_with_linenum.tsv anno_with_linenum.tsv<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>




<p>2 以构建拟南芥（Arabidopsis thaliana）的注释所需文件为例<br>第一步：准备<br>下载Arabidopsis 的 GTF file 和 genome FASTA file，到 ‘atdb’目录下。解压文件。 地址</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">mkdir atdb                                                                                                                                                
cd atdb         
wget ftp://ftp.ensemblgenomes.org/pub/release-27/plants/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.27.dna.genome.fa.gz                    
wget ftp://ftp.ensemblgenomes.org/pub/release-27/plants/gtf/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.27.gtf.gz                                                                                                                                                                                                                   
gunzip Arabidopsis_thaliana.TAIR10.27.dna.genome.fa.gz             
gunzip Arabidopsis_thaliana.TAIR10.27.gtf.gz  <span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p>**第二步：<br>用 gtfToGenePred 工具将 GTF file 转换 GenePred file</p>
<p>gtfToGenePred -genePredExt Arabidopsis_thaliana.TAIR10.27.gtf AT_refGene.txt<br>用retrieve_seq_from_fasta.pl生成 transcript FASTA file</p>
<p>perl ../retrieve_seq_from_fasta.pl –format refGene –seqfile Arabidopsis_thaliana.TAIR10.27.dna.genome.fa AT_refGene.txt AT_refGeneMrna.fa<br>After this step, the annotation database files needed for gene-based annotation are ready. Now you can annotate a given VCF file using the procedure starting from B(iii). Please note that the ‘–buildver’ argument should be set to ‘AT’.<br>参考<a target="_blank" rel="noopener" href="http://annovar.openbioinformatics.org/en/latest/user-guide/gene/">http://annovar.openbioinformatics.org/en/latest/user-guide/gene/</a> for more details.bases and other arguments are the same as in the human genome annotation.</p>
<p>总脚本如下</p>
<p>下载物种基因序列、注释文件</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">wget -c ftp://ftp.ensemblgenomes.org/pub/release-27/plants/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.27.dna.genome.fa.gz
wget -c ftp://ftp.ensemblgenomes.org/pub/release-27/plants/gtf/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.27.gtf.gz
gzip -d Arabidopsis_thaliana.TAIR10.27.dna.genome.fa.gz
gzip -d Arabidopsis_thaliana.TAIR10.27.gtf.gz<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span></span></code></pre>


<p>#gtf文件格式转换</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">wget -c http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/gtfToGenePred
gtfToGenePred -genePredExt Arabidopsis_thaliana.TAIR10.27.gtf AT_refGene.txt
 
# 另一种格式转换方法，https://github.com/chengcz/pyGTF

# 使用软件包提供脚本build物种数据库，数据库buildver为AT，名称为refGene
perl retrieve_seq_from_fasta.pl --format refGene --seqfile Arabidopsis_thaliana.TAIR10.27.dna.genome.fa AT_refGene.txt --out AT_refGeneMrna.fa<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<h2 id="五、批量注释"><a href="#五、批量注释" class="headerlink" title="五、批量注释"></a>五、批量注释</h2><p>/home/jmzeng/bio-soft/annovar/convert2annovar.pl -format vcf4  Sample3.varscan.snp.vcf &gt; Sample3.annovar<br>/home/jmzeng/bio-soft/annovar/convert2annovar.pl -format vcf4  Sample4.varscan.snp.vcf &gt; Sample4.annovar<br>/home/jmzeng/bio-soft/annovar/convert2annovar.pl -format vcf4  Sample5.varscan.snp.vcf &gt; Sample5.annovar<br>然后用下面这个脚本批量注释</p>
<p>image001<br>Reading gene annotation from /home/jmzeng/bio-soft/annovar/humandb/hg19_refGene.txt … Done with 50914 transcripts (including 11516 without coding sequence annotation) for 26271 unique genes</p>
<p>最后查看结果可知，真正在外显子上面的突变并不多</p>
<p>23515 Sample3.anno.exonic_variant_function<br>23913 Sample4.anno.exonic_variant_function<br>24009 Sample5.anno.exonic_variant_function<br>annovar软件就是把我们得到的十万多个snp分类了，看看这些snp分别是基因的哪些位置，是否引起蛋白突变</p>
<p>downstream<br>exonic<br>exonic;splicing<br>intergenic<br>intronic<br>ncRNA_exonic<br>ncRNA_intronic<br>ncRNA_splicing<br>ncRNA_UTR3<br>ncRNA_UTR5<br>splicing<br>upstream<br>upstream;downstream<br>UTR3<br>UTR5<br>UTR5;UTR3<br>六、一步到位：table_annovar.pl 可以同时进行3种注释<br>perl table_annovar.pl example/ex1.avinput humandb/ -buildver hg19 -out myanno -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a -operation gx,r,f,f,f -nastring . -csvout -polish -xref example/gene_xref.txt<br>#-remove: remove all temporary files<br>#-operation:g,gene-based; gx,gene-based with cross-reference annotation (from -xref argument);r, region-based; f,filter-based.<br>#-nastring：没有对应注释，则输出<code>.</code><br>#-csvout:结果用,分隔；去掉则采用默认，用Tab分隔<br>#-xref: whether a known genetic disease is caused by defects in this gene (this information was suffplied in the  example/gene_xref.txt file in the command line) 这一项没有也OK</p>
<p>其中（每种数据库对应的类型参考官网）<br>g,gene-based,对应数据库为refGene,ensGene等<br>r,region-based,对应数据库为cytoBand等<br>f,filter-based,对应数据库为exac03,avsnp147,dbnsfp30a等</p>
<p>3种注释分开进行：annotate_variation.pl<br>1 gene-based<br>perl annotate_variation.pl -geneanno -dbtype refGene -buildver hg19 example/ex1.avinput humandb/  </p>
<p>结果文件在example/中，ex1.avinput.variant_function和ex1.avinput.exonic_variant_function<br>（1）ex1.avinput.variant_function</p>
<p>[root@localhost example]# head ex1.avinput.variant_function<br>UTR5    ISG15(NM_005101:c.-33T&gt;C)       1       948921  948921  T       C       comments: rs15842, a SNP in 5’ UTR of ISG15<br>UTR3    ATAD3C(NM_001039211:c.*91G&gt;T)   1       1404001 1404001 G       T       comments: rs149123833, a SNP in 3’ UTR of ATAD3C</p>
<p>第1列：variant effects,将变异分类，如intergenic, intronic, non-synonymous SNP, frameshift deletion, large-scale duplication等<br>第2列：基因名，Symbol，括号中为NM_22222,为refGene编号<br>其余列：输入文件ex1.avinput的内容</p>
<p>（2）ex1.avinput.exonic_variant_function</p>
<p>[root@localhost example]# head  ex1.avinput.exonic_variant_function<br>line9   nonsynonymous SNV       IL23R:NM_144701:exon9:c.G1142A:p.R381Q, 1       67705958        67705958       GA       comments: rs11209026 (R381Q), a SNP in IL23R associated with Crohn’s disease<br>line10  nonsynonymous SNV       ATG16L1:NM_017974:exon8:c.A841G:p.T281A,ATG16L1:NM_001190267:exon9:c.A550G:p.T184A,ATG16L1:NM_030803:exon9:c.A898G:p.T300A,ATG16L1:NM_001190266:exon9:c.A646G:p.T216A,ATG16L1:NM_198890:exon5:c.A409G:p.T137A,  2       234183368       234183368       A       G       comments: rs2241880 (T300A), a SNP in the ATG16L1 associated with Crohn’s disease<br>line11  nonsynonymous SNV       NOD2:NM_022162:exon4:c.C2104T:p.R702W,NOD2:NM_001293557:exon3:c.C2023T:p.R675W,16       50745926        50745926        C       T       comments: rs2066844 (R702W), a non-synonymous SNP in NOD2</p>
<p>第1列：该变异在input文件的行号<br>第2列：对编码基因的影响，frameshift,nonsynonymous等<br>第3列：被影响的基因或转录本,其中NM_22222,为refGene编号<br>其余列：同输入文件</p>
<p>用awk操作时，分隔符设定为\t；不设置时，空格也被当做分隔符，会造成错位</p>
<p>[root@localhost example]# head  ex1.avinput.exonic_variant_function|awk -F ‘\t’ ‘{print $2}’<br>nonsynonymous SNV<br>nonsynonymous SNV<br>nonsynonymous SNV<br>nonsynonymous SNV<br>frameshift insertion<br>frameshift deletion<br>frameshift deletion<br>stoploss<br>stopgain<br>frameshift substitution</p>
<p>[root@localhost example]# head  ex1.avinput.exonic_variant_function|awk ‘{print $2}’<br>nonsynonymous<br>nonsynonymous<br>nonsynonymous<br>nonsynonymous<br>frameshift<br>frameshift<br>frameshift<br>stoploss<br>stopgain<br>frameshift</p>
<p>2 region-based<br>perl annotate_variation.pl -regionanno -dbtype cytoBand -buildver hg19 example/ex1.avinput humandb/ </p>
<p>鉴定各变异的cytogenetic band,如1p36.33<br>结果文件在example中，ex1.avinput.hg19_cytoBand</p>
<p>[root@localhost example]# more ex1.avinput.hg19_cytoBand<br>cytoBand        1p36.33 1       948921  948921  T       C       comments: rs15842, a SNP in 5’ UTR of ISG15<br>cytoBand        1p36.33 1       1404001 1404001 G       T       comments: rs149123833, a SNP in 3’ UTR of ATAD3C<br>cytoBand        1p36.31 1       5935162 5935162 A       T       comments: rs1287637, a splice site variant in NP<br>HP4<br>cytoBand        1q23.3  1       162736463       162736463       C       T       comments: rs1000050, a SNP in Il<br>lumina SNP arrays<br>第1列：cytoBand<br>第2列：1p21.1<br>其余列：同输入文件</p>
<p>3 ilter<br>perl annotate_variation.pl -filter -dbtype exac03 -buildver hg19 example/ex1.avinput humandb/<br>结果文件在example/中,ex1.avinput.hg19_exac03_filtered(exac03中没有报道的位点）和ex1.avinput.hg19_exac03_dropped（exac03中报道的位点，包含其等位基因频率）</p>
<hr>
<p>比较两个文件的重复值，保留后者重复项</p>
<pre class="line-numbers language-markup" data-language="markup"><code class="language-markup">awk 'BEGIN {OFS="\t"} 
     NR==FNR {if(FNR&gt;1) start_values[$3] = 1; next} 
     FNR==1 {print; next} 
     $2 in start_values' annovar/merged_result.tsv filtered_variants_thr0p7_20250916_203336.tsv &gt; filtered_filtered_variants_thr0p7_20250916_203336.tsv<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span></span></code></pre>



<hr>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/84c818207240">https://www.jianshu.com/p/84c818207240</a><br><a target="_blank" rel="noopener" href="https://www.plob.org/article/9976.html">https://www.plob.org/article/9976.html</a></p>
<p>snpEff<br><a target="_blank" rel="noopener" href="https://www.bioinfo-scrounger.com/archives/268/">https://www.bioinfo-scrounger.com/archives/268/</a></p>
<p>重要<br><a target="_blank" rel="noopener" href="http://www.bio-info-trainee.com/2028.html">http://www.bio-info-trainee.com/2028.html</a><br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/cea3b179b8a9">https://www.jianshu.com/p/cea3b179b8a9</a></p>
<p>ANNOVAR注释变异VCF结果说明<br><a target="_blank" rel="noopener" href="https://www.omicsclass.com/article/464">https://www.omicsclass.com/article/464</a><br><a target="_blank" rel="noopener" href="https://www.omicsclass.com/article/804">https://www.omicsclass.com/article/804</a><br><a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog_71df25810102ybtt.html">http://blog.sina.com.cn/s/blog_71df25810102ybtt.html</a></p>
<p>Hui Y, Kai W. Genomic variant annotation and prioritization with ANNOVAR and wANNOVAR[J]. Nature Protocols, 2015, 10(10).</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>Author: </span><span class="post-copyright-info"><a href="https://0HHHHHH0.github.io">HHHH</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>Link: </span><span class="post-copyright-info"><a href="https://0hhhhhh0.github.io/2025/09/16/53/">https://0hhhhhh0.github.io/2025/09/16/53/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>Copyright Notice: </span><span class="post-copyright-info">All articles on this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless otherwise stated.</span></div></div><div class="tag_share"><div class="post-share"><div class="social-share" data-image="/uploads/avatar.png" data-sites="facebook,x,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2025/09/17/55/" title="基因结构"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info"><div class="info-1"><div class="info-item-1">Previous</div><div class="info-item-2">基因结构</div></div><div class="info-2"><div class="info-item-1">结构图!!!     flowchart TD     A["一个完整的真核生物基因（DNA模板链）"]          subgraph A_Sub[基因结构]         direction LR         Promoter[启动子 Promoter]         Exon1[外显子1 Exon 1]         Intron[内含子 Intron]         Exon2[外显子2 Exon 2]         Terminator[终止子 Terminator]     end          A --&gt; A_Sub          subgraph Transcription[第一步：转录]         direction TB         B["初级mRNA转录本（Pre-mRNA, 包含所有序列）"]     end          A_Sub --转录--&gt; Transcription          subgraph Splicing[第二步：加工（剪切与拼接）]         direction TB  ...</div></div></div></a><a class="pagination-related" href="/2025/09/03/50/" title="Typecho快速修改文章图片和文件链接地址"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">Next</div><div class="info-item-2">Typecho快速修改文章图片和文件链接地址</div></div><div class="info-2"><div class="info-item-1">说明为了使我们的博客速度更快，图片等一些文件会采用外接的形式。但当跟换域名或CDN图片地址时，需要大批修正一切文章的援用图片链接和一些文件链接地址，手动修正又烦劳，这里就说一下通过 phpMyAdmin工具快速修正数据库文章图片，文件链接地址。  1.安装如果您使用的是基于 Debian 或 Ubuntu 的 Linux 系统，可以尝试使用 apt-get 命令。例如：  sudo apt-get updatesudo apt-get install phpmyadmin   The phpmyadmin package must have a database installed and configured before it cwith dbconfig-common. If you are an advanced database administrator and know that you want to performdatabase has already been installed and configured, you should refuse this ...</div></div></div></a></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/uploads/avatar.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">HHHH</div><div class="author-info-description">总有人间一两风，填我十万八千梦</div><div class="site-data"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">16</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">1</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">8</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/0HHHHHH0"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons"><a class="social-icon" href="https://github.com/0hhhhhh0" target="_blank" title="fab fa-github"><i class="GitHub"></i></a><a class="social-icon" href="mailto:695460776@qq.com" target="_blank" title="fa fa-envelope"><i class="E-Mail"></i></a><a class="social-icon" href="https://x.com/your-user-name" target="_blank" title="fab fa-twitter"><i class="Twitter"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">随便逛逛!</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%80%E3%80%81%E7%AE%80%E4%BB%8B"><span class="toc-number">1.</span> <span class="toc-text">一、简介</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BA%8C%E3%80%81ANNOVAR%E7%9A%84%E4%B8%8B%E8%BD%BD"><span class="toc-number">2.</span> <span class="toc-text">二、ANNOVAR的下载</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#ANNOVAR-%E7%9A%84%E6%96%87%E4%BB%B6%E7%BB%93%E6%9E%84"><span class="toc-number">2.1.</span> <span class="toc-text">ANNOVAR 的文件结构</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%89%E3%80%81%E6%9E%84%E5%BB%BA%E8%87%AA%E5%AE%9A%E4%B9%89%E6%B3%A8%E9%87%8A%E5%BA%93"><span class="toc-number">3.</span> <span class="toc-text">三、构建自定义注释库</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%AC%AC%E4%B8%80%E6%AD%A5%EF%BC%9A%E5%87%86%E5%A4%87%E5%B7%A5%E4%BD%9C"><span class="toc-number">3.1.</span> <span class="toc-text">第一步：准备工作</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%AC%AC%E4%BA%8C%E6%AD%A5%EF%BC%9A%E5%BB%BA%E7%AB%8B%E6%B3%A8%E9%87%8A%E5%BA%93"><span class="toc-number">3.2.</span> <span class="toc-text">第二步：建立注释库</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%AC%AC%E4%B8%89%E6%AD%A5%EF%BC%9A%E8%BD%AC%E6%8D%A2%E9%9C%80%E8%A6%81%E6%B3%A8%E9%87%8A%E7%9A%84vcf%E6%96%87%E4%BB%B6"><span class="toc-number">3.3.</span> <span class="toc-text">第三步：转换需要注释的vcf文件</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%AC%AC%E5%9B%9B%E6%AD%A5%EF%BC%9Aannotate-variation%E6%B3%A8%E9%87%8A"><span class="toc-number">4.</span> <span class="toc-text">第四步：annotate_variation注释</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BA%94%E3%80%81%E6%89%B9%E9%87%8F%E6%B3%A8%E9%87%8A"><span class="toc-number">5.</span> <span class="toc-text">五、批量注释</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Posts</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/10/09/hexo-butterfly-ge-ren-bo-ke-da-jian/" title="Hexo + Butterfly 个人博客搭建">Hexo + Butterfly 个人博客搭建</a><time datetime="2025-10-09T14:45:00.000Z" title="Created 2025-10-09 22:45:00">2025-10-09</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/56/" title="410随身WiFi棒子Debian刷入青龙教程">410随身WiFi棒子Debian刷入青龙教程</a><time datetime="2025-09-25T11:26:00.000Z" title="Created 2025-09-25 19:26:00">2025-09-25</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/17/55/" title="基因结构">基因结构</a><time datetime="2025-09-17T14:24:00.000Z" title="Created 2025-09-17 22:24:00">2025-09-17</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/16/53/" title="ANNOVAR | 变异注释">ANNOVAR | 变异注释</a><time datetime="2025-09-16T13:43:00.000Z" title="Created 2025-09-16 21:43:00">2025-09-16</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/03/50/" title="Typecho快速修改文章图片和文件链接地址">Typecho快速修改文章图片和文件链接地址</a><time datetime="2025-09-03T07:20:00.000Z" title="Created 2025-09-03 15:20:00">2025-09-03</time></div></div></div></div></div></div></main><footer id="footer"><div class="footer-other"><div class="footer-copyright"><span class="copyright">&copy;&nbsp;2025 By HHHH</span><span class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 7.3.0</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.5.1</a></span></div></div><div class="container" id="jsi-flying-fish-container"><script src="js/fish.js"></script></div><style>   @media only screen and (max-width: 767px){
   #sidebar_search_box input[type=text]{width:calc(100% - 24px)}
}</style></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Reading Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light and Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle Between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Settings"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back to Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=5.5.1"></script><script src="/js/main.js?v=5.5.1"></script><div class="js-pjax"></div><script type="text/javascript" src="/js/jquery.min.js"></script><script type="text/javascript" src="/js/fish.js" defer data-pjax></script><script type="text/javascript" src="/js/busuanzi.pure.mini.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">Search</span><i class="fas fa-spinner fa-pulse" id="loading-status" hidden="hidden"></i><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="text-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  Loading Database</span></div><div class="local-search-input"><input placeholder="Search for Posts" type="text"/></div><hr/><div id="local-search-results"></div><div class="ais-Pagination" id="local-search-pagination" style="display:none;"><ul class="ais-Pagination-list"></ul></div><div id="local-search-stats"></div></div><div id="search-mask"></div><script src="/js/search/local-search.js?v=5.5.1"></script></div></div></body></html>